library/src/main/scala/scala/scalajs/concurrent/QueueExecutionContext.scala

```diff
@@ -1,22 +1,55 @@
 package scala.scalajs.concurrent
 
 import scala.concurrent.ExecutionContextExecutor
-import scalajs.js
 
-private[concurrent] object QueueExecutionContext
-    extends ExecutionContextExecutor {
+import scala.scalajs.js
+import scala.scalajs.js.|
 
-  def execute(runnable: Runnable): Unit = {
-    js.timers.setTimeout(0) {
-      try {
-        runnable.run()
-      } catch {
-        case t: Throwable => reportFailure(t)
-      }
+object QueueExecutionContext {
+  def timeouts(): ExecutionContextExecutor =
+    new TimeoutsExecutionContext
+
+  def promises(): ExecutionContextExecutor =
+    new PromisesExecutionContext
+
+  def apply(): ExecutionContextExecutor =
+    if (js.isUndefined(js.Dynamic.global.Promise)) timeouts()
+    else promises()
+
+  private final class TimeoutsExecutionContext extends ExecutionContextExecutor {
+    def execute(runnable: Runnable): Unit = {
+      js.Dynamic.global.setTimeout({ () =>
```

Comment:
This is not consistent with `PromisesExecutionContext` as it doesn't capture `setTimeout`. I don't know how much sense it makes to do so, but maybe it is a good idea for consistency.